sdlib/d/gc/tcache.d

```diff
@@ -54,6 +64,77 @@ public:
 		pd.arena.free(emap, pd, ptr);
 	}
 
+	/**
+	 * Appendables API.
+	 */
+
+	// Get the capacity of the array slice denoted by slice and length.
+	// It includes the space occupied by the slice itself.
+	size_t getArrayCapacity(void* sliceAddr, size_t sliceLen) {
+		if ((sliceAddr is null) || (!sliceLen))
+			return 0;
+
+		auto pd = maybeGetPageDescriptor(sliceAddr);
+		// Return zero if slice is unknown to GC or is to a non-appendable block:
+		if ((pd.extent is null) || (!pd.extent.allocSize))
+			return 0;
+
+		// If an element exists after the slice, capacity is defined as zero.
+		// See also: https://dlang.org/spec/arrays.html#capacity-reserve
+		if (sliceAddr - pd.extent.address + sliceLen < pd.extent.allocSize)
+			return 0;
+
+		// Otherwise, capacity is the block size minus the slice length:
+		return pd.extent.size - sliceLen;
```

Comment:
This is incorrect, the slice doesn't have to start at the beginning of the extent.

Reply:
This is clearly wrong. Remove everything else, write the proper tests, and it'll be obvious in no time.